### DISCUSSION TITLE
How to handle auto reconnect on device disconnected?

### DISCUSSION THREAD

**Discussion by juanaza** [capacitor-community/bluetooth-le#420] [Q&A]
Hi guys, I would like to elaborate on an issue I am currently having with this plugin and mainly on the little knowledge I have on BLE communication. I have been working with this plugin some months, and I haven't arrived yet to the smartest/optimal solution to handle reconnect after device disconnects from phone. 

I will paste some screenshots of my current implementation:
1) Phone scans for devices with a name prefix and a service

2) Once I have the list of scanned devices and I connect to one, I do the following:
- RequestLeScan to make sure BLE device is still available and then connect to the device (taking into consideration the existing issue in Android "disconnect before connect")
- Register a onDisconnect event that will handle the reconnect process
```
async connectDevice(device: string) {
    await BleClient.requestLEScan(
      {
        namePrefix: "DeviceName",
        optionalServices: [this.deviceCustomServiceUUID]
      },
      async (result: ScanResult) => {
        if (device === result.device.deviceId) {
          await BleClient.stopLEScan();
          await BleClient.disconnect(device);

          await BleClient.connect(device,
            async (device) => {
              await this.onDeviceDisconnected(device);
            });

          await this.onDeviceConnected(device);
        }
      }
    );

    setTimeout(async () => {
      await BleClient.stopLEScan();
    }, 3000);
  }

private async onDeviceConnected(device: string) {
    this.spbDevice = device;

    if (this.reconnectInterval) {
      clearInterval(this.reconnectInterval);
    }

    await this.stopAllNotifications();
    await this.startAllNotifications();

    if (this.sessionDetail.isTraining == false) {
      this.sessionDetail.state = SpbDeviceState.CONNECTED;
    }
    else {
      this.sessionDetail.state = SpbDeviceState.TRAINING;
      await this.switchSpbState(SpbDeviceState.TRAINING);
    }
  }
```

3) Once device disconnects the previous onDeviceDisconnect functionality fires and I manually try to connect to my device every 10 seconds with the use of an interval. Once the device reconnects successfully I clear this interval.

```
private async onDeviceDisconnected(device: string) {    
    if (this.sessionDetail.isTraining) {
      this.sessionDetail.state = DeviceState.TRAINING_DISCONNECTED;

      this.reconnectInterval = setInterval(async () => {
        let isPhoneReady = await this.isPhoneReadyForDeviceConnection();

        if (isPhoneReady) {
          this.connectSpb(device);
        }
      }, 10000);
    }
    else {
      this.sessionDetail.state = DeviceState.DISCONNECTED;
    }
    this.sessionDetailSub.next(this.sessionDetail);
  }
```

Apart from my doubt regarding handling the reconnect phase after a disconnect, is it correct to stop and start notifications before a connection taking into account I also handle an automatic reconnect? Meaning that stop/start notifications could be triggered multiple times. I have realized that if notifications are not started again after a reconnect, I am not receiving data from my subscriptions eventhough they are correctly sent from the device.

**Comment by pwespi** [comment]
Thanks for the discussion.

Regarding reconnection, it is hard to give general advice. I think it really depends on the use case of your app. In some cases you might want to have some logic to automatically reconnect, like you're describing. In other cases you might just want to inform the user and the user decides what to do.

Regarding notifications, yes, you should start them again after a new connection.
